File: companion-ios/Sources/GPXExporter.swift
import Foundation
import CoreLocation

// Serializes a route polyline to GPX 1.1 <trk> the device's parser reads
// (it scans lat="…"/lon="…" attribute pairs, order lat-before-lon).
enum GPXExporter {
    static func make(name: String, coordinates: [CLLocationCoordinate2D]) -> String {
        var s = """
        <?xml version="1.0" encoding="UTF-8"?>
        <gpx version="1.1" creator="OpenCycleInk" xmlns="http://www.topografix.com/GPX/1/1">
        <trk><name>\(xmlEscape(name))</name><trkseg>
        """
        s.reserveCapacity(coordinates.count * 48)
        for c in coordinates {
            s += String(format: "<trkpt lat=\"%.6f\" lon=\"%.6f\"></trkpt>",
                        c.latitude, c.longitude)
        }
        s += "</trkseg></trk></gpx>"
        return s
    }

    private static func xmlEscape(_ s: String) -> String {
        s.replacingOccurrences(of: "&", with: "&amp;")
            .replacingOccurrences(of: "<", with: "&lt;")
            .replacingOccurrences(of: ">", with: "&gt;")
    }
}

File: companion-android/app/src/main/java/com/raemond/opentrailpaper/data/GpxExporter.kt
package com.raemond.opentrailpaper.data

import java.util.Locale

/**
 * Serializes a route polyline to the GPX 1.1 `<trk>` the device's parser reads
 * (it scans lat="…"/lon="…" attribute pairs, order lat-before-lon).
 *
 * [Locale.US] on every format call, not the default: a phone set to a
 * comma-decimal locale would otherwise emit lat="37,7749", which the device's
 * scanner reads as 37 — a route that lands a hundred kilometres away.
 */
object GpxExporter {
    fun make(name: String, coordinates: List<LatLon>): String {
        val s = StringBuilder(coordinates.size * 48 + 256)
        s.append("<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n")
        s.append("<gpx version=\"1.1\" creator=\"OpenTrailPaper\" ")
        s.append("xmlns=\"http://www.topografix.com/GPX/1/1\">\n")
        s.append("<trk><name>").append(xmlEscape(name)).append("</name><trkseg>\n")
        for (c in coordinates) {
            s.append(
                String.format(
                    Locale.US,
                    "<trkpt lat=\"%.6f\" lon=\"%.6f\"></trkpt>",
                    c.lat,
                    c.lon,
                ),
            )
        }
        s.append("</trkseg></trk></gpx>")
        return s.toString()
    }

    private fun xmlEscape(s: String) =
        s.replace("&", "&amp;").replace("<", "&lt;").replace(">", "&gt;")
}
